src/nominatim_db/tools/special_phrases/sp_importer.py

```diff
@@ -177,10 +232,17 @@ def _create_classtype_table_and_indexes(self,
         with self.db_connection.cursor() as db_cursor:
             db_cursor.execute("CREATE INDEX idx_placex_classtype ON placex (class, type)")
 
+        allowed_special_phrases = self.get_sp()
```

Comment:
You don't really need to get the intersection of database and style class/types here. It is enough to fetch the class/types from the database. The rest of the code takes care of reading from the style and the actual intersection happens then through the condition in line 241.